### PR TITLE
using empty namespace for types with no package

### PR DESCRIPTION
**Pull request by ashkboos**
As suggested in issue #138 we now use empty namespace in the FastenUris of types without package instead of using null

### CODE REVIEW

File: analyzer/javacg-opal/src/test/java/eu/fasten/analyzer/javacgopal/CGandStitchingTest.java [hunk 1]
```diff
@@ -102,6 +103,12 @@ public void staticInitializer() throws OPALException, URISyntaxException {
             "\n", toString(mergedRcg));
     }
 
+    @Test
+    public void noPackage() throws URISyntaxException {
+        final var np = getRCG("merge/staticInitializer/NoPackage.class", "noPackage", "0.0.0");
```
Comment (proksch) [inline analyzer/javacg-opal/src/test/java/eu/fasten/analyzer/javacgopal/CGandStitchingTest.java:108]: Shouldn't this be "NoPackage"? (capital case N)

File: analyzer/javacg-opal/src/main/java/eu/fasten/analyzer/javacgopal/data/analysis/OPALMethod.java [hunk 2]
```diff
@@ -104,7 +107,7 @@ public static String getPackageName(final Type parameter) {
                 return slashToDot(Objects
                         .requireNonNull(getPackageName(parameter.asArrayType().componentType())));
             } else if (parameter.asObjectType().packageName().equals("")) {
-                return null;
+                return "";
```
Comment (proksch) [inline analyzer/javacg-opal/src/main/java/eu/fasten/analyzer/javacgopal/data/analysis/OPALMethod.java:110]: Do we then end up with a double slash in the URLs? So before it was "/null/ClassName", now it is "//ClassName"... or is this being handled through canonicalize?

**Review by proksch — COMMENTED** [review]
I have two minor clarification questions...

File: analyzer/javacg-opal/src/test/java/eu/fasten/analyzer/javacgopal/CGandStitchingTest.java (hunk 1 above)
Comment (ashkboos) [reply analyzer/javacg-opal/src/test/java/eu/fasten/analyzer/javacgopal/CGandStitchingTest.java]: fixed

File: analyzer/javacg-opal/src/main/java/eu/fasten/analyzer/javacgopal/data/analysis/OPALMethod.java (hunk 2 above)
Comment (ashkboos) [reply analyzer/javacg-opal/src/main/java/eu/fasten/analyzer/javacgopal/data/analysis/OPALMethod.java:110]: Yes, now it is "//ClassName". Canonicalize does not handle it. In fact, if there is no namespace canonicalize complains about it. That is why we now check before canonicalizing to see if there is not any namespace please don't try.

Comment (proksch) [reply analyzer/javacg-opal/src/main/java/eu/fasten/analyzer/javacgopal/data/analysis/OPALMethod.java:110]: Hmm, I do not like the `//` construct... I would argue that the correct version would be `/ClassName`. Is there an easy way to achieve this?

Comment (ashkboos) [reply analyzer/javacg-opal/src/main/java/eu/fasten/analyzer/javacgopal/data/analysis/OPALMethod.java:110]: It is not allowd by the FastenUri class if you write this: 
```
var v = FastenURI.create("/ClassName");
```
you would get `java.lang.IllegalArgumentException: Missing entity`

Comment (ashkboos) [reply analyzer/javacg-opal/src/main/java/eu/fasten/analyzer/javacgopal/data/analysis/OPALMethod.java:110]: I also just realized that even "//ClassName" cannot be correctly parsed by FastenUri. It does not cause any exception but It assumes that `ClassName` is the product name while it is the name of the class. What should we do about this? should we use some meaningful placeholder for the package name such as `/no.package/ClassName`? or should we create another URI specific for the classes?

Comment (proksch) [reply analyzer/javacg-opal/src/main/java/eu/fasten/analyzer/javacgopal/data/analysis/OPALMethod.java:110]: How about using `/ClassName` and fixing the code that extracts the package/product/class information?

Comment (proksch) [reply analyzer/javacg-opal/src/main/java/eu/fasten/analyzer/javacgopal/data/analysis/OPALMethod.java:110]: In general, I think we should revise the `FastenURI` at some point. We should improve our validation and I also think that it does not make sense to parse the whole thing on initialization. I would prefer storing the string representation and only parse part of it on demand, we can still cache information, once parsed. ... but this is a discussion for another issue.

**Review by proksch — CHANGES_REQUESTED** [review]
It seems that we should introduce proper handling for the problem at hand as an ad-hoc solution does not seem to solve the problem. I propose that we move this to the next release.

**Comment by mir-am** [comment]
With Mehdi, we decided to close this PR. Because we should do a major refactoring on the FASTEN URIs or use another technique.

**Comment by amottier** [comment]
@mir-am @ashkboos did you open an issue to track discussion on the "major refactoring on the FASTEN URIs" mentioned in the @mir-am comment? I guess that a refactoring would address issue #138, #248 and #260. Also can you remind me in which deliverable current implementation of URI is specified? Thanks.

**Pull request closed without merging**